compiler/rustc_passes/src/check_attr.rs

```diff
@@ -352,7 +352,13 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
     }
 
     /// Checks if `#[diagnostic::do_not_recommend]` is applied on a trait impl.
-    fn check_do_not_recommend(&self, attr_span: Span, hir_id: HirId, target: Target) {
+    fn check_do_not_recommend(
+        &self,
+        attr_span: Span,
+        hir_id: HirId,
+        target: Target,
+        attr: &Attribute,
+    ) {
         if !matches!(target, Target::Impl) {
```

Comment:
Should this also be checking that the target is a trait impl and not inherent?
Otherwise it seems to accept `#[diagnostic::do_not_recommend]` on `impl T {}`

Reply:
Good catch. I've pushed https://github.com/rust-lang/rust/pull/132056/commits/3d814b3b562d01d33160046d406dc6473a6f5455 to fix this.